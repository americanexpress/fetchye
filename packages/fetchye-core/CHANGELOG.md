# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [1.1.0](https://github.com/americanexpress/fetchye/compare/v1.0.0...v1.1.0) (2021-10-07)


### Bug Fixes

* **defaultFetcher:** Avoid known hang in node-fetch ([d33399c](https://github.com/americanexpress/fetchye/commit/d33399c71c6809a868c8f6b928ed349d995d8f6a))
* **jsonError:** update to handle NO CONTENT response ([030f886](https://github.com/americanexpress/fetchye/commit/030f8863de37fef253fcac477d8f30c55bd88cf7))
* **makeServerFetchye:** mask errors from server side fetching ([#49](https://github.com/americanexpress/fetchye/issues/49)) ([3503836](https://github.com/americanexpress/fetchye/commit/3503836ea914ff4877af1682c7c672e834286264))





## [1.0.4](https://github.com/americanexpress/fetchye/compare/v1.0.0...v1.0.4) (2021-08-10)


### Bug Fixes

* **defaultFetcher:** Avoid known hang in node-fetch ([d33399c](https://github.com/americanexpress/fetchye/commit/d33399c71c6809a868c8f6b928ed349d995d8f6a))
* **jsonError:** update to handle NO CONTENT response ([030f886](https://github.com/americanexpress/fetchye/commit/030f8863de37fef253fcac477d8f30c55bd88cf7))
* **makeServerFetchye:** mask errors from server side fetching ([#49](https://github.com/americanexpress/fetchye/issues/49)) ([3503836](https://github.com/americanexpress/fetchye/commit/3503836ea914ff4877af1682c7c672e834286264))





## [1.0.3](https://github.com/americanexpress/fetchye/compare/v1.0.0...v1.0.3) (2021-07-14)


### Bug Fixes

* **defaultFetcher:** Avoid known hang in node-fetch ([d33399c](https://github.com/americanexpress/fetchye/commit/d33399c71c6809a868c8f6b928ed349d995d8f6a))
* **jsonError:** update to handle NO CONTENT response ([030f886](https://github.com/americanexpress/fetchye/commit/030f8863de37fef253fcac477d8f30c55bd88cf7))





## [1.0.2](https://github.com/americanexpress/fetchye/compare/v1.0.0...v1.0.2) (2021-03-08)


### Bug Fixes

* **jsonError:** update to handle NO CONTENT response ([030f886](https://github.com/americanexpress/fetchye/commit/030f8863de37fef253fcac477d8f30c55bd88cf7))





## 1.0.1 (2020-11-17)


### Features

* **fetchye:** prepare for initial release [skip ci] ([#28](https://github.com/americanexpress/fetchye/issues/28)) ([2b212df](https://github.com/americanexpress/fetchye/commit/2b212df8fab4405e2b7c51ad687a280cfe27ebbd))





# 1.0.0 (2020-11-12)


### Features

* **fetchye:** prepare for initial release [skip ci] ([#28](https://github.com/americanexpress/fetchye/issues/28)) ([2b212df](https://github.com/americanexpress/fetchye/commit/2b212df8fab4405e2b7c51ad687a280cfe27ebbd))





# [1.0.0-beta.6](https://github.com/americanexpress/fetchye/compare/v1.0.0-beta.5...v1.0.0-beta.6) (2020-11-12)

**Note:** Version bump only for package fetchye-core





# [1.0.0-beta.5](https://github.com/americanexpress/fetchye/compare/v1.0.0-beta.3...v1.0.0-beta.5) (2020-11-04)



# 1.0.0-beta.4 (2020-11-03)

**Note:** Version bump only for package fetchye-core





# [1.0.0-beta.4](https://github.com/americanexpress/fetchye/compare/v1.0.0-beta.3...v1.0.0-beta.4) (2020-11-03)

**Note:** Version bump only for package fetchye-core
