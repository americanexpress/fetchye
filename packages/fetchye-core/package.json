{
  "name": "fetchye-core",
  "version": "1.0.0-beta.4",
  "description": "Core components of the Fetchye library",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "module": "es/index.js",
  "files": [
    "lib",
    "es"
  ],
  "scripts": {
    "clean": "rimraf lib && rimraf es",
    "prebuild": "yarn clean",
    "build:cjs": "cross-env BABEL_ENV=commonjs babel --extensions .js,.jsx ./src --out-dir lib",
    "build:esm": "babel --extensions .js,.jsx ./src --out-dir es",
    "build": "yarn build:cjs && yarn build:esm"
  },
  "keywords": [
    "fetchye-core",
    "fetchye"
  ],
  "contributors": [
    "James Singleton <James.Singleton1@aexp.com> https://github.com/JamesSingleton)",
    "Jimmy King  <Jimmy.King@aexp.com> (https://github.com/10xLaCroixDrinker)",
    "Jonathan Adshead <Jonathan.Adshead@aexp.com> (https://github.com/JAdshead)",
    "Michael Tobia <Michael.M.Tobia@aexp.com> (https://github.com/Francois-Esquire)",
    "Nelly Kiboi <Nelly.J.Kiboi@aexp.com> (https://github.com/nellyk)",
    "Ruben Casas <ruben.casas@aexp.com> (https://github.com/infoxicator)"
  ],
  "dependencies": {
    "create-shared-react-context": "^1.0.3"
  },
  "devDependencies": {
    "@babel/cli": "^7.12.1",
    "@babel/core": "^7.12.3",
    "@testing-library/react-hooks": "^3.4.2",
    "babel-preset-amex": "^3.4.1",
    "cross-env": "^7.0.2",
    "react-test-renderer": "^16.14.0",
    "rimraf": "^3.0.2"
  },
  "peerDependencies": {
    "react": "^16.14.0",
    "react-dom": "^16.14.0"
  }
}
