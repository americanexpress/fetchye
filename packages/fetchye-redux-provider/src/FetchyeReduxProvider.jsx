/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, {
  useMemo, useRef, useEffect, useReducer, useCallback,
} from 'react';
import PropTypes from 'prop-types';
import { useDispatch, useStore } from 'react-redux';
import {
  defaultFetcher, FetchyeContext, defaultEqualityChecker, useSubscription,

} from 'fetchye-core';

const makeUseFetchyeSelector = ({
  getCacheByKey, cacheSelector, equalityChecker, subscribe,
}) => (key) => {
  const [, forceRender] = useReducer((n) => n + 1, 0);
  const store = useStore();
  const selector = useCallback((state) => getCacheByKey(cacheSelector(state), key), [key]);
  const initialValue = selector(store.getState());

  const lastSelectorValue = useRef(initialValue);
  const selectorValue = useRef(initialValue);

  useEffect(() => {
    function checkForUpdates() {
      const nextValue = selector(store.getState());
      lastSelectorValue.current = selectorValue.current;
      selectorValue.current = nextValue;

      if (!equalityChecker(selectorValue.current, lastSelectorValue.current)) {
        forceRender();
      }
    }
    checkForUpdates();
    return subscribe(checkForUpdates);
  }, [selector, store]);

  return selectorValue;
};

const FetchyeReduxProvider = ({
  cache,
  fetcher = defaultFetcher,
  equalityChecker = defaultEqualityChecker,
  fetchClient = fetch,
  children,
}) => {
  const [notify, subscribe] = useSubscription();
  const reduxDispatch = useDispatch();
  const { cacheSelector, getCacheByKey } = cache;

  const memoizedConfig = useMemo(() => ({
    cache,
    dispatch: (...args) => {
      reduxDispatch(...args);
      notify();
    },
    defaultFetcher: fetcher,
    useFetchyeSelector: makeUseFetchyeSelector({
      getCacheByKey,
      cacheSelector,
      equalityChecker,
      subscribe,
    }),
    fetchClient,
  }), [
    cache,
    cacheSelector,
    fetchClient,
    fetcher,
    getCacheByKey,
    equalityChecker,
    reduxDispatch,
    notify,
    subscribe,
  ]);
  return (
    <FetchyeContext.Provider value={memoizedConfig}>
      {children}
    </FetchyeContext.Provider>
  );
};

FetchyeReduxProvider.propTypes = {
  cache: PropTypes.shape({
    reducer: PropTypes.func.isRequired,
    getCacheByKey: PropTypes.func.isRequired,
    cacheSelector: PropTypes.func.isRequired,
  }).isRequired,
  equalityChecker: PropTypes.func,
  fetchClient: PropTypes.func,
  fetcher: PropTypes.func,
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
};

export default FetchyeReduxProvider;
