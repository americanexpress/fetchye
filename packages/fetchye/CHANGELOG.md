# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [1.4.0](https://github.com/americanexpress/fetchye/compare/v1.3.0...v1.4.0) (2023-10-23)


### Features

* **fetchye:** support headers as a function ([#86](https://github.com/americanexpress/fetchye/issues/86)) ([cd4c175](https://github.com/americanexpress/fetchye/commit/cd4c175d87c1391d155a70834ef98798251f9134))
* **mapOptionToKey:** provide first helper ([9202e22](https://github.com/americanexpress/fetchye/commit/9202e226a48c4bab87024cc0175907fad03c4fbb))





# [1.3.0](https://github.com/americanexpress/fetchye/compare/v1.2.0...v1.3.0) (2023-08-28)


### Features

* **signal:** remove passed signal from defautlMapOptionsToKey ([#83](https://github.com/americanexpress/fetchye/issues/83)) ([8bb4b36](https://github.com/americanexpress/fetchye/commit/8bb4b36647d6004277153362526434fc25f230bd))





## [1.2.1](https://github.com/americanexpress/fetchye/compare/v1.2.0...v1.2.1) (2023-04-06)

**Note:** Version bump only for package fetchye





# [1.2.0](https://github.com/americanexpress/fetchye/compare/v1.1.0...v1.2.0) (2023-03-07)


### Bug Fixes

* **cache:** remove previous data or error on new run ([#66](https://github.com/americanexpress/fetchye/issues/66)) ([f921ef3](https://github.com/americanexpress/fetchye/commit/f921ef301c7ac28172772acccae0f6274cf76225))
* **peerDeps:** creating fix commit for versioning reasons ([#60](https://github.com/americanexpress/fetchye/issues/60)) ([f63a28e](https://github.com/americanexpress/fetchye/commit/f63a28e09b0796870c5d8b8388ebbdc4bd8639f9))


### Features

* **compute-key:** add support for mapKeyToCacheKey ([#71](https://github.com/americanexpress/fetchye/issues/71)) ([1f29629](https://github.com/americanexpress/fetchye/commit/1f29629a8d4e053546f08c089739f205ee7bbc05))





## [1.1.2](https://github.com/americanexpress/fetchye/compare/v1.1.0...v1.1.2) (2022-10-21)


### Bug Fixes

* **cache:** remove previous data or error on new run ([#66](https://github.com/americanexpress/fetchye/issues/66)) ([f921ef3](https://github.com/americanexpress/fetchye/commit/f921ef301c7ac28172772acccae0f6274cf76225))
* **peerDeps:** creating fix commit for versioning reasons ([#60](https://github.com/americanexpress/fetchye/issues/60)) ([f63a28e](https://github.com/americanexpress/fetchye/commit/f63a28e09b0796870c5d8b8388ebbdc4bd8639f9))





## [1.1.1](https://github.com/americanexpress/fetchye/compare/v1.1.0...v1.1.1) (2022-07-05)


### Bug Fixes

* **peerDeps:** creating fix commit for versioning reasons ([#60](https://github.com/americanexpress/fetchye/issues/60)) ([f63a28e](https://github.com/americanexpress/fetchye/commit/f63a28e09b0796870c5d8b8388ebbdc4bd8639f9))





# [1.1.0](https://github.com/americanexpress/fetchye/compare/v1.0.0...v1.1.0) (2021-10-07)


### Bug Fixes

* **defaultFetcher:** Avoid known hang in node-fetch ([d33399c](https://github.com/americanexpress/fetchye/commit/d33399c71c6809a868c8f6b928ed349d995d8f6a))
* **FetchyeProvider:** correct stale data bug ([#47](https://github.com/americanexpress/fetchye/issues/47)) ([2ff8c5a](https://github.com/americanexpress/fetchye/commit/2ff8c5a46c0266852d010cdf800adfe84f6d8d42))
* **jsonError:** update to handle NO CONTENT response ([030f886](https://github.com/americanexpress/fetchye/commit/030f8863de37fef253fcac477d8f30c55bd88cf7))
* **makeServerFetchye:** mask errors from server side fetching ([#49](https://github.com/americanexpress/fetchye/issues/49)) ([3503836](https://github.com/americanexpress/fetchye/commit/3503836ea914ff4877af1682c7c672e834286264))


### Features

* **fetchye:** http headers casing does not affect cache key creation ([#53](https://github.com/americanexpress/fetchye/issues/53)) ([5ad1dc2](https://github.com/americanexpress/fetchye/commit/5ad1dc202b35453b5cf036a3792e52e4046be94f))





## [1.0.4](https://github.com/americanexpress/fetchye/compare/v1.0.0...v1.0.4) (2021-08-10)


### Bug Fixes

* **defaultFetcher:** Avoid known hang in node-fetch ([d33399c](https://github.com/americanexpress/fetchye/commit/d33399c71c6809a868c8f6b928ed349d995d8f6a))
* **FetchyeProvider:** correct stale data bug ([#47](https://github.com/americanexpress/fetchye/issues/47)) ([2ff8c5a](https://github.com/americanexpress/fetchye/commit/2ff8c5a46c0266852d010cdf800adfe84f6d8d42))
* **jsonError:** update to handle NO CONTENT response ([030f886](https://github.com/americanexpress/fetchye/commit/030f8863de37fef253fcac477d8f30c55bd88cf7))
* **makeServerFetchye:** mask errors from server side fetching ([#49](https://github.com/americanexpress/fetchye/issues/49)) ([3503836](https://github.com/americanexpress/fetchye/commit/3503836ea914ff4877af1682c7c672e834286264))





## [1.0.3](https://github.com/americanexpress/fetchye/compare/v1.0.0...v1.0.3) (2021-07-14)


### Bug Fixes

* **defaultFetcher:** Avoid known hang in node-fetch ([d33399c](https://github.com/americanexpress/fetchye/commit/d33399c71c6809a868c8f6b928ed349d995d8f6a))
* **jsonError:** update to handle NO CONTENT response ([030f886](https://github.com/americanexpress/fetchye/commit/030f8863de37fef253fcac477d8f30c55bd88cf7))





## [1.0.2](https://github.com/americanexpress/fetchye/compare/v1.0.0...v1.0.2) (2021-03-08)


### Bug Fixes

* **jsonError:** update to handle NO CONTENT response ([030f886](https://github.com/americanexpress/fetchye/commit/030f8863de37fef253fcac477d8f30c55bd88cf7))





## 1.0.1 (2020-11-17)


### Features

* **fetchye:** prepare for initial release [skip ci] ([#28](https://github.com/americanexpress/fetchye/issues/28)) ([2b212df](https://github.com/americanexpress/fetchye/commit/2b212df8fab4405e2b7c51ad687a280cfe27ebbd))





# 1.0.0 (2020-11-12)


### Features

* **fetchye:** prepare for initial release [skip ci] ([#28](https://github.com/americanexpress/fetchye/issues/28)) ([2b212df](https://github.com/americanexpress/fetchye/commit/2b212df8fab4405e2b7c51ad687a280cfe27ebbd))





# [1.0.0-beta.6](https://github.com/americanexpress/fetchye/compare/v1.0.0-beta.5...v1.0.0-beta.6) (2020-11-12)

**Note:** Version bump only for package fetchye





# [1.0.0-beta.5](https://github.com/americanexpress/fetchye/compare/v1.0.0-beta.3...v1.0.0-beta.5) (2020-11-04)



# 1.0.0-beta.4 (2020-11-03)

**Note:** Version bump only for package fetchye





# [1.0.0-beta.4](https://github.com/americanexpress/fetchye/compare/v1.0.0-beta.3...v1.0.0-beta.4) (2020-11-03)

**Note:** Version bump only for package fetchye
