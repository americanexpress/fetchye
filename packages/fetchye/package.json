{
  "name": "fetchye",
  "version": "1.1.0",
  "description": "If you know how to use Fetch, you know how to use Fetchye [fetch-yae]. Simple React Hooks, Centralized Cache, Infinitely Extensible.",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "module": "es/index.js",
  "files": [
    "lib",
    "es"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/americanexpress/fetchye.git",
    "directory": "packages/fetchye"
  },
  "bugs": {
    "url": "https://github.com/americanexpress/fetchye/issues"
  },
  "homepage": "https://github.com/americanexpress/fetchye#readme",
  "scripts": {
    "clean": "rimraf lib && rimraf es",
    "prebuild": "yarn clean",
    "build:cjs": "cross-env BABEL_ENV=commonjs babel --extensions .js,.jsx ./src --out-dir lib",
    "build:esm": "babel --extensions .js,.jsx ./src --out-dir es",
    "build": "yarn build:cjs && yarn build:esm"
  },
  "keywords": [
    "async",
    "react",
    "redux",
    "react-redux",
    "hooks",
    "ssr",
    "fetch",
    "data",
    "query",
    "promise"
  ],
  "author": "Michael A Tomcal <michael.a.tomcal@aexp.com>",
  "contributors": [
    "James Singleton <James.Singleton1@aexp.com> https://github.com/JamesSingleton)",
    "Jamie King  <Jamie.King@aexp.com> (https://github.com/10xLaCroixDrinker)",
    "Jonathan Adshead <Jonathan.Adshead@aexp.com> (https://github.com/JAdshead)",
    "Michael Tobia <Michael.M.Tobia@aexp.com> (https://github.com/Francois-Esquire)",
    "Michael Tomcal <Michael.A.Tomcal@aexp.com> (https://github.com/mtomcal))",
    "Nelly Kiboi <Nelly.J.Kiboi@aexp.com> (https://github.com/nellyk)",
    "Nickolas Oliver <nickolas.oliver@aexp.com> (https://github.com/PixnBits)",
    "Andrew Curtis <andrew.curtis@aexp.com> (https://github.com/drewcur)",
    "Scott McIntyre <scott.mcintyre@aexp.com> (https://github.com/smackfu)",
    "Ruben Casas <ruben.casas@aexp.com> (https://github.com/infoxicator)"
  ],
  "dependencies": {
    "@babel/runtime": "^7.11.2",
    "fetchye-core": "^1.1.0",
    "object-hash": "^2.0.3",
    "prop-types": "^15.7.2"
  },
  "peerDependencies": {
    "react": ">=16.14.0",
    "react-dom": ">=16.14.0"
  },
  "devDependencies": {
    "@babel/cli": "^7.12.1",
    "@babel/core": "7.11.6",
    "@testing-library/react": "^11.0.4",
    "babel-preset-amex": "^3.4.1",
    "cross-env": "^7.0.2",
    "fetchye-test-utils": "^1.1.0",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-redux": "^7.2.1",
    "redux": "^4.0.5",
    "rimraf": "^3.0.2"
  }
}
