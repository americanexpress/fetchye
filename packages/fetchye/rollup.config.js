import { nodeResolve } from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import babel from '@rollup/plugin-babel';
import { terser } from 'rollup-plugin-terser';
import pkg from './package.json';

const inputSrc = 'src/index.js';

export default [
  {
    input: inputSrc,
    output: {
      name: 'fetchye',
      file: pkg.browser,
      format: 'umd',
    },
    external: ['react', 'react-dom'],
    plugins: [
      nodeResolve({ extensions: ['.js', '.jsx'], browser: true }),
      commonjs(),
      babel({
        babelHelpers: 'bundled',
        exclude: 'node_modules/**',
        presets: ['amex'],
      }),
      terser(),
    ],
  },
  {
    input: inputSrc,
    output: {
      file: pkg.main,
      format: 'cjs',
    },
    external: ['react', 'react-dom'],
    plugins: [
      nodeResolve({ extensions: ['.js', '.jsx'] }),
      commonjs(),
      babel({
        babelHelpers: 'runtime',
        exclude: 'node_modules/**',
        presets: ['amex'],
        plugins: [['@babel/plugin-transform-runtime', { useESModules: false }]],
      }),
      terser(),
    ],
  },
  {
    input: inputSrc,
    output: {
      file: pkg.module,
      format: 'es',
    },
    external: ['react', 'react-dom'],
    plugins: [
      nodeResolve({ extensions: ['.js', '.jsx'] }),
      commonjs(),
      babel({
        babelHelpers: 'runtime',
        exclude: 'node_modules/**',
        presets: ['amex'],
        plugins: [['@babel/plugin-transform-runtime', { useESModules: true }]],
      }),
      terser(),
    ],
  },
];
