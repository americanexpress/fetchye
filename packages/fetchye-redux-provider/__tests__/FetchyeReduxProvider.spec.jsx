/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { useContext } from 'react';
import { act, render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { FetchyeContext } from 'fetchye-core';
import { SimpleCache } from 'fetchye';
import PropTypes from 'prop-types';
import FetchyeReduxProvider from '../src/FetchyeReduxProvider';

const store = createStore((state) => state, { initialState: {} });

global.fetch = () => {};

describe('FetchyeReduxProvider', () => {
  it('should create fetchye context with global fetch', () => {
    let contextResult;
    render(
      <Provider store={store}>
        <FetchyeReduxProvider cache={SimpleCache()}>
          {React.createElement(() => {
            contextResult = useContext(FetchyeContext);
            return <fake-element />;
          })}
        </FetchyeReduxProvider>
      </Provider>
    );
    expect(contextResult).toMatchInlineSnapshot(`
      Object {
        "cache": Object {
          "cacheSelector": [Function],
          "getCacheByKey": [Function],
          "reducer": [Function],
        },
        "defaultFetcher": [Function],
        "dispatch": [Function],
        "fetchClient": [Function],
        "useFetchyeSelector": [Function],
      }
    `);
  });
  it('should create fetchye context with custom fetch', () => {
    const fakeFetchClient = jest.fn();
    render(
      <Provider store={store}>
        <FetchyeReduxProvider cache={SimpleCache()} fetchClient={fakeFetchClient}>
          {React.createElement(() => {
            const { fetchClient } = useContext(FetchyeContext);
            fetchClient();
            return <fake-element />;
          })}
        </FetchyeReduxProvider>
      </Provider>
    );
    expect(fakeFetchClient).toHaveBeenCalled();
  });
  it('should call cacheSelector within useFetchyeSelector', () => {
    const fakeCacheSelector = jest.fn((state) => state);
    const cache = SimpleCache({ cacheSelector: fakeCacheSelector });
    render(
      <Provider store={store}>
        <FetchyeReduxProvider cache={cache}>
          {React.createElement(() => {
            const { useFetchyeSelector } = useContext(FetchyeContext);
            useFetchyeSelector();
            return <fake-element />;
          })}
        </FetchyeReduxProvider>
      </Provider>
    );
    expect(fakeCacheSelector).toHaveBeenCalled();
  });
  it('should return a stable response from useFetchyeSelector that changes properly with a changed input', () => {
    const fakeCacheSelector = jest.fn(() => ({
      data: {
        key1: 'val1',
        key2: 'val2',
      },
      loading: {},
      errors: {},
    }));
    const cache = SimpleCache({ cacheSelector: fakeCacheSelector });
    const captureValue = jest.fn();

    const Component = ({ id }) => {
      const { useFetchyeSelector } = useContext(FetchyeContext);
      const selectedRef = useFetchyeSelector(id);
      captureValue(selectedRef.current);
      return <fake-element />;
    };

    Component.propTypes = {
      id: PropTypes.string.isRequired,
    };

    const { rerender } = render(
      <Provider store={store}>
        <FetchyeReduxProvider cache={cache}>
          <Component id="key1" />
        </FetchyeReduxProvider>
      </Provider>
    );
    act(() => {
      rerender(
        <Provider store={store}>
          <FetchyeReduxProvider cache={cache}>
            <Component id="key2" />
          </FetchyeReduxProvider>
        </Provider>
      );
    });

    // The value will have been captured 3 times.
    // The first render causes val1 to be captured.
    // The second render causes val1 to be captured again, but queues the effect
    // The effect causes a third render that captures val2 as expected
    expect(captureValue).toMatchSnapshot();
  });
});
