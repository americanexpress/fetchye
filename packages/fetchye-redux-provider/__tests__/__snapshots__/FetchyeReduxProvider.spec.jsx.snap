// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchyeReduxProvider should have loading state change when data is populated 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": undefined,
        "error": undefined,
        "loading": true,
        "query": undefined,
      },
    ],
    Array [
      Object {
        "data": "val1",
        "error": undefined,
        "loading": false,
        "query": undefined,
      },
    ],
    Array [
      Object {
        "data": undefined,
        "error": undefined,
        "loading": false,
        "query": undefined,
      },
    ],
    Array [
      Object {
        "data": undefined,
        "error": undefined,
        "loading": false,
        "query": undefined,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`FetchyeReduxProvider should return a stable response from useFetchyeSelector that changes properly with a changed input 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "val1",
        "error": undefined,
        "loading": false,
        "query": undefined,
      },
    ],
    Array [
      Object {
        "data": "val1",
        "error": undefined,
        "loading": false,
        "query": undefined,
      },
    ],
    Array [
      Object {
        "data": "val2",
        "error": undefined,
        "loading": false,
        "query": undefined,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
